### PR TITLE
Loading & Error Screens

### PR DESCRIPTION
**Pull request by thisjustin123**
## Overview:

Adds Loading and Error screens for the main page with full UI implementation and functionality. I wish I could demo the app loading correctly, but backend is still down 🙃 

## Changes:
- New `MainLoading` subscreen.
- New `MainError` subscreen.
- Refactor of `UpliftApiRepository` to allow `MainError` to try to load backend again.
- More documentation added to API code.

## Demos:

<details>
<summary>Loading/Error/Reloading Demo:</summary>
<br>

The API request takes like 30 seconds to crash, feel free to skip through to find the error part of the demo @ 00:44.

https://github.com/cuappdev/uplift-android/assets/47724806/ca2ff095-13e0-4f27-b6c4-677fa3055dc5

</details>

### CODE REVIEW

File: app/src/main/java/com/cornellappdev/uplift/ui/viewmodels/HomeViewModel.kt [hunk 1]
```diff
@@ -57,20 +54,22 @@ class HomeViewModel : ViewModel() {
...
+                        -1
+                    } else if (!isCurrentlyOpen(gym1.hours[todayIndex()]) && isCurrentlyOpen(
+                            gym2.hours[todayIndex()]
+                        )
+                    ) {
+                        1
+                    }
+                    // Both are either favorited and unfavorited, too.
+                    else {
+                        // TODO: Convert to compare based off distance.
```
Comment (emilyhu8) [inline app/src/main/java/com/cornellappdev/uplift/ui/viewmodels/HomeViewModel.kt:69]: for distance do you plan on using the eatery method of 2 coordinates?

File: app/src/main/java/com/cornellappdev/uplift/ui/screens/subscreens/MainLoading.kt [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                                cornerRadius = 8.dp,
+                                paddingValues = PaddingValues(end = 16.dp)
+                            )
+                        }
+                    }
+                }
+            }
+        }
+
+        // Fake section to make it look better :-)
```
Comment (emilyhu8) [inline app/src/main/java/com/cornellappdev/uplift/ui/screens/subscreens/MainLoading.kt:78]: LOL

File: app/src/main/java/com/cornellappdev/uplift/ui/components/general/UpliftTopBar.kt [hunk 3]
```diff
@@ -54,4 +69,31 @@ fun UpliftTopBar(
...
+            modifier = Modifier
+                .fillMaxWidth()
+                .align(Alignment.Bottom),
+            verticalAlignment = Alignment.Bottom
+        ) {
+            Surface(
+                shape = RoundedCornerShape(10.dp),
+                color = Color.White,
+                modifier = Modifier.size(width = 186.dp, height = 20.dp)
+            ) {}
```
Comment (emilyhu8) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/general/UpliftTopBar.kt:95]: can delete the extra curly brackets {} I think?

**Review by emilyhu8 — APPROVED** [review]
looks great!

File: app/src/main/java/com/cornellappdev/uplift/ui/components/general/UpliftTopBar.kt (hunk 3 above)
Comment (thisjustin123) [reply app/src/main/java/com/cornellappdev/uplift/ui/components/general/UpliftTopBar.kt:95]: Can't delete the content parameter for `Surface` unfortunately... not sure why, but that means I need the `{}` to compile.

File: app/src/main/java/com/cornellappdev/uplift/ui/viewmodels/HomeViewModel.kt (hunk 1 above)
Comment (thisjustin123) [reply app/src/main/java/com/cornellappdev/uplift/ui/viewmodels/HomeViewModel.kt:69]: Yeah, I think I will. I don't think we're displaying walking distances, so we can just use a general estimate like coordinates here and it should be fine. (If anything it's more questionable that Eatery is using coordinates to compute walking distances heh)

**Pull request merged**